File: src/main/java/dev/kensa/KensaExecutionContext.java
package dev.kensa;

import dev.kensa.context.TestContainer;
import dev.kensa.output.ResultWriter;
import org.junit.jupiter.api.extension.ExtensionContext;

import java.util.*;

import static java.util.Collections.synchronizedList;

public class KensaExecutionContext implements ExtensionContext.Store.CloseableResource {

    private final ResultWriter resultWriter;
    private final List<TestContainer> containers = synchronizedList(new ArrayList<>());

    KensaExecutionContext(ResultWriter resultWriter) {
        this.resultWriter = resultWriter;
    }

    @Override
    public void close() {
        resultWriter.write(sortedContainers());
    }

    void register(TestContainer testContainer) {
        containers.add(testContainer);
    }

    private Set<TestContainer> sortedContainers() {
        SortedSet<TestContainer> sortedContainers = new TreeSet<>(Comparator.comparing(o -> o.testClass().getName()));
        sortedContainers.addAll(containers);

        return sortedContainers;
    }
}

File: src/main/java/dev/kensa/KensaExecutionContext.kt
package dev.kensa

import dev.kensa.context.TestContainer
import java.util.*
import java.util.Collections.synchronizedList

class KensaExecutionContext {
    private val _containers = synchronizedList(ArrayList<TestContainer>())
    val containers: List<TestContainer> get() = _containers

    fun add(testContainer: TestContainer) {
        _containers.add(testContainer)
    }
}